app/src/org/commcare/android/database/app/models/FormDefRecord.java

```diff
@@ -0,0 +1,183 @@
+
+    @Persisting(4)
+    @MetaField(META_FORM_FILE_PATH)
+    private String mFormFilePath;
+
+    @Persisting(value = 6, nullable = true)
+    @MetaField(META_BASE64_RSA_PUBLIC_KEY)
+    private String mBase64RsaPublicKey;
+
+    @Persisting(11)
+    @MetaField(META_FORM_MEDIA_PATH)
+    private String mFormMediaPath;
+
+    @Persisting(value = 12, nullable = true)
+    @MetaField(META_MODEL_VERSION)
+    private int mModelVersion = -1;
+
+    @Persisting(value = 13, nullable = true)
+    @MetaField(META_UI_VERSION)
+    private int mUiVersion = -1;
+
+    //    Serialization Only!
+    public FormDefRecord() {
+    }
+
+    public FormDefRecord(String displayName, String jrFormId, String formFilePath, String formMediaPath) {
+        mDisplayName = displayName;
+        mJrFormId = jrFormId;
+        mFormFilePath = formFilePath;
+        mFormMediaPath = formMediaPath;
+    }
+
+    // Only for DB Migration
+    public FormDefRecord(Cursor cursor) {
+        mDisplayName = cursor.getString(cursor.getColumnIndex(FormsProviderAPI.FormsColumns.DISPLAY_NAME));
+        mJrFormId = cursor.getString(cursor.getColumnIndex(FormsProviderAPI.FormsColumns.JR_FORM_ID));
+        mModelVersion = cursor.getInt(cursor.getColumnIndex(FormsProviderAPI.FormsColumns.MODEL_VERSION));
+        mUiVersion = cursor.getInt(cursor.getColumnIndex(FormsProviderAPI.FormsColumns.UI_VERSION));
+        mFormMediaPath = cursor.getString(cursor.getColumnIndex(FormsProviderAPI.FormsColumns.FORM_MEDIA_PATH));
+        mFormFilePath = cursor.getString(cursor.getColumnIndex(FormsProviderAPI.FormsColumns.FORM_FILE_PATH));
+        mBase64RsaPublicKey = cursor.getString(cursor.getColumnIndex(FormsProviderAPI.FormsColumns.BASE64_RSA_PUBLIC_KEY));
+    }
+
+    public static Vector<Integer> getFormDefIdsByJrFormId(SqlStorage<FormDefRecord> formDefRecordStorage, String jrFormId) {
+        return formDefRecordStorage.getIDsForValue(META_JR_FORM_ID, jrFormId);
+    }
+
+    public static Vector<FormDefRecord> getFormDefsByJrFormId(SqlStorage<FormDefRecord> formDefRecordStorage, String jrFormId) {
+        return formDefRecordStorage.getRecordsForValue(META_JR_FORM_ID, jrFormId);
+    }
+
+    public int save(SqlStorage<FormDefRecord> formDefRecordStorage) {
+        // if we don't have a path to the file, the rest are irrelevant.
+        // it should fail anyway because you can't have a null file path.
+        if (StringUtils.isEmpty(mFormFilePath)) {
+            throw new SQLException("Can't save a form def with an empty form file path");
+        }
+        // Make sure that the necessary fields are all set
+
+        File form = new File(mFormFilePath);
+        if (StringUtils.isEmpty(mDisplayName)) {
+            mDisplayName = form.getName();
+        }
+
+        if (StringUtils.isEmpty(mFormMediaPath)) {
+            mFormMediaPath = getMediaPath(mFormFilePath);
+        }
+
+        formDefRecordStorage.write(this);
+
+        if (recordId == -1) {
```

Comment:
It can definitely return -1 which means the insertion eventually failed due to unknown reasons. I think we are safe though as there is not anything stateful that happens in SqlStorage.write().

Reply:
Cool, gotcha. Killing this thread then